### ISSUE TITLE
Error when feeding tf.keras model in custom loss function

### ISSUE THREAD

**Issue by batzner** [tensorflow/tensorflow#23769]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux openSUSE Leap 42.3
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: none
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v1.12.0-0-ga6d8ffae09 1.12.0
- Python version: 3.4.6
- Bazel version (if compiling from source): none
- GCC/Compiler version (if compiling from source): none
- CUDA/cuDNN version: none
- GPU model and memory: none

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**

I am trying to add adversarial training to my `tf.keras` model. For this, I added a custom loss function, which feeds the model the adversarially perturbed input and adds the cross-entropy to the loss. 

The problem is that feeding the model a tensor in the custom loss function leads to a `TypeError: argument of type 'NoDependency' is not iterable`. I have reduced it to a minimal example that simply feeds the model zeros below. 

If I call `model(model.input)` before the custom loss function is called, it surprisingly works.

**Describe the expected behavior**

The expected behavior would be the model compiling and fitting as it does with the `model(model.input)` line uncommented. The example below would also work when defining the `y_pred_zeros = model(tf.zeros((32, 20)))` outside of the `loss` function. But for adversarial training, I need to call the model inside of the loss function.

**Code to reproduce the issue**
```python
import numpy as np
import tensorflow as tf

model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(20,)),
    tf.keras.layers.Dense(5)
])

# With the following line uncommented, it works.
# model(model.input)

def loss(y_true, y_pred):
    loss = tf.keras.losses.mean_squared_error(y_true, y_pred)

    # Feed zeros to the model and add the mean of the output to the loss
    y_pred_zeros = model(tf.zeros((32, 20)))

    return loss + tf.reduce_mean(y_pred_zeros)

model.compile('sgd', loss=loss)

x_train, y_train = np.random.randn(100, 20), np.random.randn(100, 5)
model.fit(x_train, y_train)
```

**Other info / logs**
```python
/home/kilian/Desktop/venv/bin/python /home/kilian/Desktop/tmp.py
Traceback (most recent call last):
  File "/home/kilian/Desktop/tmp.py", line 20, in <module>
    model.compile('sgd', loss=loss)
  File "/home/kilian/Desktop/venv/lib/python3.4/site-packages/tensorflow/python/training/checkpointable/base.py", line 474, in _method_wrapper
    method(self, *args, **kwargs)
  File "/home/kilian/Desktop/venv/lib/python3.4/site-packages/tensorflow/python/keras/engine/training.py", line 617, in compile
    output_loss = weighted_loss(y_true, y_pred, sample_weight, mask)
  File "/home/kilian/Desktop/venv/lib/python3.4/site-packages/tensorflow/python/keras/engine/training_utils.py", line 598, in weighted
    score_array = fn(y_true, y_pred)
  File "/home/kilian/Desktop/tmp.py", line 16, in loss
    y_pred_zeros = model(tf.zeros((32, 20)))
  File "/home/kilian/Desktop/venv/lib/python3.4/site-packages/tensorflow/python/keras/engine/base_layer.py", line 704, in __call__
    if ('mask' in self._call_fn_args and 'mask' not in kwargs and
TypeError: argument of type 'NoDependency' is not iterable

Process finished with exit code 1
```

**Comment by harshini-gadige** [comment]
@fchollet  Tried at my end and could reproduce the same issue. Please take a look.

**Comment by hadim** [comment]
I confirm the code to reproduce the issue works on the latest TF 2 nightly build (as of today). You can close here.

**Comment by ymodak** [comment]
This is fixed in latest tf-nightly version 1.15.0-dev20190725. Thanks!
